icar/others.py

```diff
@@ -0,0 +1,69 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import List, Optional
+
+from pydantic import BaseModel, Field
+
+from . import enums, types
+
+
+class UnitCode(Enum):
+    SEC = "SEC"
+    MIN = "MIN"
+
+
+class SupportedMessage(BaseModel):
+    messages: Optional[enums.IcarMessageType] = None
+
+
+class PlanOrActual(Enum):
+    Plan = "Plan"
+    Actual = "Actual"
+
+
+class UnitCode2(Enum):
+    KGM = "KGM"
+
+
+class IcarQuarterId(Enum):
+    LF = "LF"
+    RF = "RF"
+    LR = "LR"
+    RR = "RR"
+
+
+class UnitCode3(Enum):
+    KGM = "KGM"
+    LBR = "LBR"
+
+
+class Fraction(BaseModel):
+    breed: Optional[types.IcarBreedIdentifierType] = Field(
+        None, description="The breed for this breed fraction using a scheme and id."
+    )
+    fraction: Optional[float] = Field(
+        None, description="The proportion of the denominator that this breed comprises."
+    )
+
+
+class HeatReportScrSenseTime(BaseModel):
+    breedingWindow: Optional[int] = Field(None, description="Number of hours to AI.")
+    heatIndex: Optional[int] = Field(
+        None, description="Gives an indication of the certainty of the heat indication."
+    )
+
+
+class HeatReportNedapCowControl(BaseModel):
+    expirationDateTime: Optional[types.IcarDateTimeType] = Field(
+        None,
+        description="RFC3339 UTC date/time when the heat will end (see https://ijmacd.github.io/rfc3339-iso8601/ for format guidance).",
+    )
+    heatChance: Optional[int] = Field(
+        None, description="Gives an indication of the certainty of the heat indication."
+    )
+
+
+class VisualDetection(BaseModel):
+    heatSigns: Optional[List[enums.IcarReproHeatSignType]] = None
+    heatIntensity: Optional[enums.IcarReproHeatIntensityType] = None
```

Comment:
@cookeac 

I have `collections`, `enums`, `others`, `resources` and `types` based on the folders of the specifications.

I added specials rules to to classify the remaining into modules:

- `View` and `BatchResults`, arrays go to `collections`
- Geometry Point, etc. went to "shapes" (it could have been resources too).

Everything else go to `others.py`, I would prefer to move all the classes in other more specific modules to get rid of `others.py`.

Do you have any recommendation on how we should group the remaining classes in different modules?

You can see the rules in `get_destination_module` of `scripts/main.py`.

Reply:
I agree, it is reasonable to put `View` and `BatchResults` arrays into collections.py, and the geometry features into geojson.py (as it is the GeoJSON specification which is referenced/imported).

The two HeatReport objects are deprecated. They could stay in others.py or be put into a deprecated_objects file. Unfortunately I don't see an easy way to tidy up what were in-line enumerations. My preference would be to tidy these up in a future version of the ICAR ADE specification, which would allow these to be moved to enums.py.